### PR TITLE
Add OpenAPI generation command and Ruby API step

### PR DESCRIPTION
**Pull request by JealousGx**
Introduce a command to generate OpenAPI specifications for the project and integrate Python virtual environment activation. Add a Ruby API step with its configuration and handler, along with tests to ensure the correct generation of the OpenAPI JSON file. Update the .gitignore to exclude generated files and fix minor issues in the OpenAPI title and validation.

### CODE REVIEW

**Comment by rohitg00** [comment]
Hey @JealousGx! This is an amazing work. We are planning to implement this, and your help is much appreciated. Looking forward to your PR once ready. Thanks!

File: packages/docs/content/docs/development-guide/cli.mdx [hunk 1]
```diff
@@ -127,6 +127,13 @@ Options:
 
 - `-d, --dir <step file path>`: The path relative to the steps directory to create the step file.
 
+#### `generate openapi`
+
+Generate OpenAPI spec for your project.
+
+```bash
+npx motia generate openapi
+```
```
Comment (sergiofilhowz) [inline packages/docs/content/docs/development-guide/cli.mdx:136]: can you add the list of options available?

i noticed you added title and version but we should also add the output file

File: packages/e2e/tests/release/cli/cli-validation.spec.ts [hunk 2]
```diff
@@ -60,6 +60,26 @@ test.describe('CLI Validation', () => {
...
+      })
+    } catch (error) {
+      console.error('Error generating openapi.json:', error)
+      expect(error).toBeNull()
+    }
+
+    expect(existsSync(openapiPath)).toBeTruthy()
+
+    expect(readFileSync(openapiPath, 'utf8')).toMatchSnapshot('openapi.json')
+  })
```
Comment (sergiofilhowz) [inline packages/e2e/tests/release/cli/cli-validation.spec.ts:81]: this test is good but we also need a test to validate the logic

File: packages/snap/src/cli.ts [hunk 3]
```diff
@@ -161,6 +161,17 @@ generate
     })
   })
 
+generate
+  .command('openapi')
+  .description('Generate OpenAPI spec for your project')
+  .option('-t, --title <title>', 'Title for the OpenAPI document')
+  .option('-v, --version <version>', 'Version for the OpenAPI document', '1.0.0')
```
Comment (sergiofilhowz) [inline packages/snap/src/cli.ts]: we're missing another option to select the output file path

File: packages/snap/package.json [hunk 4]
```diff
@@ -68,6 +68,7 @@
     "jest": "^29.7.0",
     "ts-jest": "^29.2.5",
     "typescript": "^5.7.2",
-    "typescript-transform-paths": "^3.5.3"
+    "typescript-transform-paths": "^3.5.3",
+    "openapi-types": "^12.1.3"
```
Comment (sergiofilhowz) [inline packages/snap/package.json:72]: we need the updated package lock file

File: packages/snap/src/generate-openapi.ts [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+  if (Array.isArray(schema.anyOf)) {
+    const nullIndex = schema.anyOf.findIndex((item: any) => item && item.type === 'null');
+
+    if (nullIndex !== -1) {
+      schema.anyOf.splice(nullIndex, 1);
+
+      schema.nullable = true;
+
+      if (schema.anyOf.length === 1) {
```
Comment (sergiofilhowz) [inline packages/snap/src/generate-openapi.ts]: what happens if it can be more than one?

File: packages/snap/src/generate-openapi.ts [hunk 6]
```diff
@@ -0,0 +1,175 @@
+
+import { LockedData } from '@motiadev/core';
+import * as fs from 'fs';
+import { OpenAPIV3 } from 'openapi-types';
+import * as path from 'path';
+import { generateLockedData } from './generate-locked-data';
+
+export async function generateOpenApi(projectDir: string, title?: string, version?: string) {
```
Comment (sergiofilhowz) [inline packages/snap/src/generate-openapi.ts]: let's move these functions to a separate folder, it will be easier to maintain later, and to write tests, we need tests for this, make sure to add test cases for some edge cases like anyof (with non nullable types) and nulllable

```
openapi/
  generate-openapi.ts
  process-schema.ts
  openapi-utils.ts
```

for tests, create a folder `__tests__/` under `src/` and add `openapi/` inside

```
src/
  __tests__/
    openapi/
      generate-openapi.spec.ts
      process-schema.spec.ts
```

File: packages/snap/src/generate-openapi.ts [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+          required: false,
+          schema: {
+            type: 'string',
+          },
+        });
+      }
+    }
+
+    if (step.config.bodySchema) {
+      const bodySchema = { ...step.config.bodySchema } as any;
```
Comment (sergiofilhowz) [inline packages/snap/src/generate-openapi.ts]: remove casting, this can lead to side-effects without TS compile complaints

File: packages/snap/src/generate-openapi.ts [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+          'application/json': {
+            schema: bodySchema,
+          },
+        },
+      };
+    }
+
+    if (step.config.responseSchema) {
+      for (const [statusCode, responseSchema] of Object.entries(step.config.responseSchema)) {
+        const resSchema = { ...responseSchema } as any;
```
Comment (sergiofilhowz) [inline packages/snap/src/generate-openapi.ts]: remove casting, this can lead to side-effects without TS compile complaints

**Review by sergiofilhowz — CHANGES_REQUESTED** [review]
hi @JealousGx  thank you so much for the PR, we wanted to add it for quite some time but never had time to work on it.

The PR looks good, but I'd like to work on a few things before we merge it.

- We need to ensure our user base are not affected by issues, that's why it's quite important to have automated tests, i saw you added one playwright test, which is good, but we need to validate inputs and outputs. We need unit tests with some json schema variants and the expected output.
- We need to remove some `any` from the code because they can lead to side-effects when we do some refactor in the future that could be hard to track down.
- Let's move this logic to a separate folder for better organization.
- Small updates for the docs.

We really appreciate your contributions and take your time to address these comments. We're excited to merge it!

File: packages/snap/src/generate-openapi.ts (hunk 5 above)
Comment (JealousGx) [reply packages/snap/src/generate-openapi.ts]: We leave them as they are. Here, we are just converting `null` to `nullable` & removing `null` from the list as per standards of OpenAPI. Others like `string`, `number` remain unchanged.

File: packages/snap/src/openapi/generate.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+import { OpenAPIV3 } from 'openapi-types';
+import * as path from 'path';
+
+import { generateLockedData } from '../generate-locked-data';
+import { processSchema } from './process-schema';
+import { isHttpMethod } from './utils';
+
+export async function generateOpenApi(projectDir: string, title?: string, version?: string, outputFile = 'openapi.json') {
+    const lockedData: LockedData = await generateLockedData(projectDir);
+    const apiSteps = lockedData.apiSteps();
```
Comment (sergiofilhowz) [inline packages/snap/src/openapi/generate.ts]: can you decouple this function from LockedData? I saw in the tests it's a little rough to read because of mocking locked data, instead, make it just return an array of API Steps

**Comment by rohitg00** [comment]
Hey @JealousGx, how's this going? Are you still working on the changes proposed by Sergio? Or you are expecting a reply. Thanks

**Comment by sergiofilhowz** [comment]
Thank you @JealousGx, sorry for not responding earlier. The code looks good! I think it's missing just formatting to run. Can you make sure it's properly formatted from prettier? Also make sure linters are fine as well

You will probably need to rebase to resolve one conflict
image

We're taking a look at the pipelines for the reason they're not running

**Comment by rohitg00** [comment]
Hey @JealousGx, can you please do the small changes Sergio requested, then it's ready to merge?
